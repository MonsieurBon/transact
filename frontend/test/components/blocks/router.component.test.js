/**
 * @jest-environment jsdom
 */

import {
  fixture,
  fixtureCleanup,
} from '@open-wc/testing-helpers/index-no-side-effects.js';
import Router from '../../../src/components/blocks/router.component.js';

class TestComponent extends HTMLElement {
  constructor() {
    super();
  }
}

describe('Router component', () => {
  let authState, router;

  beforeAll(() => {
    authState = {
      isAuthenticated: jest.fn(() => false),
    }
    router = {
      register: jest.fn(),
      deregister: jest.fn(),
      redirect: jest.fn()
    };
    window.customElements.define('t-router', Router(authState, router));
    window.customElements.define('t-test', TestComponent);
  });

  afterEach(() => {
    fixtureCleanup();
    jest.clearAllMocks();
  });

  it('should render', async () => {
    const el = await fixture('<t-router></t-router>');
    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should register itself with the router', async () => {
    const el = await fixture('<t-router></t-router>');
    expect(router.register).toHaveBeenCalledTimes(1);
    expect(router.register).toHaveBeenCalledWith(el);
  });

  it('should deregister itself from the router', async () => {
    const el = await fixture('<t-router></t-router>');
    el.parentElement.removeChild(el);

    expect(router.register).toHaveBeenCalledTimes(1);
    expect(router.deregister).toHaveBeenCalledTimes(1);
    expect(router.deregister).toHaveBeenCalledWith(el);
  });

  it('should render the current path', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><t-test route="/foo"></t-test><t-test route="/bar"></t-test></t-router>'
    );

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should replace the rendered node', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><t-test route="/foo"></t-test><t-test route="/bar"></t-test></t-router>'
    );

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();

    history.pushState({}, '', '/bar');
    el.navigationStateChanged();

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should not rerender if not necessary', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><t-test route="/foo"></t-test><t-test route="/bar"></t-test></t-router>'
    );

    expect(el.navigationStateChanged()).toBeFalsy();
  });

  it('should redirect to the default route if not found', async () => {
    history.pushState({}, '', '/bar');
    const el = await fixture(
      '<t-router default-route="/foo"><t-test route="/foo"></t-test></t-router>'
    );

    expect(router.redirect).toHaveBeenCalledWith('/foo');
  });

  it('should render a secured node if authenticated', async () => {
     authState.isAuthenticated.mockImplementationOnce(() => true);

    history.pushState({}, '', '/foo');
    const el = await fixture(
      '<t-router><t-test route="/foo" secured="true"></t-test></t-router>'
    );

    expect(el.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should only redirect once if default route is secured', async () => {
    history.pushState({}, '', '/foo');
    const el = await fixture(
      `<t-router default-route="/bar" default-route-unsecure="/baz">
         <t-test route="/foo" secured></t-test>
         <t-test route="/bar" secured></t-test>
         <t-test route="/baz"></t-test>
       </t-router>`
    );

    expect(router.redirect).toHaveBeenCalledTimes(1);
    expect(router.redirect).toHaveBeenCalledWith('/baz');
  });
});
