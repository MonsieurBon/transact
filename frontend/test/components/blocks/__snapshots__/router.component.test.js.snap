// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router component should render 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"></div>
"
`;

exports[`Router component should render a secured node if authenticated 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><t-test route=\\"/foo\\" secured=\\"true\\"></t-test></div>
"
`;

exports[`Router component should render the current path 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><t-test route=\\"/foo\\"></t-test></div>
"
`;

exports[`Router component should replace the rendered node 1`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><t-test route=\\"/foo\\"></t-test></div>
"
`;

exports[`Router component should replace the rendered node 2`] = `
"
<style>
  :host {
    contain: content;
  }
  
  slot {
    display: none;
  }
</style>
<div id=\\"router-outlet\\"><t-test route=\\"/bar\\"></t-test></div>
"
`;
